reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java

```diff
@@ -0,0 +1,404 @@
+
+        private int pending;
+        // need sync since should be checked/zerroed in onNext
+        // and increased in request
+        private int counter; // no need to sync since increased zerroed only in
+        // the request method
+
+        final CoreSubscriber<? super T> actual;
+
+        Subscription s;
+
+        private LimitRateSubscriber(CoreSubscriber<? super T> actual,
+            int highTide,
+            int lowTide) {
+            this.actual = actual;
+            this.highTide = highTide;
+            this.limit = Operators.unboundedOrLimit(highTide, lowTide);
+        }
+
+        @Override
+        public void request(long n) {
+            if (Operators.validate(n)) {
+                Operators.addCap(REQUESTED, this, n);
+
+                tryRequest();
+            }
+        }
+
+        void tryRequest() {
+            if (WIP.getAndIncrement(this) != 0) {
+                return;
+            }
+
+
+            final Subscription s = this.s;
+            final int highTide = this.highTide;
+
+            int missed = 1;
+            int pending = this.pending;
+            int produced = this.produced;
+            long r = this.requested;
+            long deliver = 0;
+
+            for (;;) {
+                if (r != Long.MAX_VALUE || highTide != Integer.MAX_VALUE) {
+                    if (produced != 0) {
+                        deliver = Math.min(produced, r);
+                    }
+                    else if (pending != highTide) {
+                        deliver = Math.min(highTide - pending, r);
+                    }
+
+                    int w = wip;
+                    if (missed == w) {
+                        if (deliver > 0) {
+                            if (produced != 0) {
+                                this.pending = highTide - produced + (int) deliver;
+                                PRODUCED.addAndGet(this, -produced);
+                            }
+                            else {
+                                this.pending = pending + (int) deliver;
+                            }
+                            Operators.addCap(REQUESTED, this, -deliver);
+
+                            missed = WIP.addAndGet(this, -missed);
+                            if (missed == 0) {
+                                s.request(deliver);
+                                break;
+                            }
+                        } else {
+                            if (produced != 0) {
+                                this.pending = highTide - produced;
+                                PRODUCED.addAndGet(this, -produced);
+                            }
+
+                            missed = WIP.addAndGet(this, -missed);
+                            if (missed == 0) {
+                                break;
+                            }
+                        }
+
+                        produced = this.produced;
```

Comment:
would it make sense to always make these two variable refresh at the beginning of the loop? (at least right after the first `if (unboundedRequest || unboundedHighTide)`)